File: scrapeless-sdks/go/conformance_test.go
package scrapeless

import (
	"encoding/json"
	"os"
	"path/filepath"
	"strings"
	"testing"
)

// Reads ../conformance.json, the shared truth for all nine SDKs, so a fold that
// drifts here fails here. Only the standard library, so `go test` runs without
// fetching chromedp or launching a browser.

type foldVector struct {
	Raw    string `json:"raw"`
	Expect string `json:"expect"`
}

type expectation struct {
	Available            bool           `json:"available"`
	Reason               string         `json:"reason"`
	Total                int            `json:"total"`
	ReportedTotal        int            `json:"reportedTotal"`
	Categories           map[string]int `json:"categories"`
	IsAntibot            bool           `json:"isAntibot"`
	IsCaptcha            bool           `json:"isCaptcha"`
	IsFingerprinted      bool           `json:"isFingerprinted"`
	IsProtected          bool           `json:"isProtected"`
	FirstName            string         `json:"firstName"`
	FirstCategoryKey     string         `json:"firstCategoryKey"`
	LastConfidenceIsNull bool           `json:"lastConfidenceIsNull"`
}

type resultVector struct {
	Name   string         `json:"name"`
	Detail map[string]any `json:"detail"`
	Expect expectation    `json:"expect"`
}

type conformance struct {
	CategoryFold []foldVector   `json:"categoryFold"`
	Results      []resultVector `json:"results"`
	Unavailable  struct {
		Expect expectation `json:"expect"`
	} `json:"unavailable"`
}

func loadConformance(t *testing.T) conformance {
	t.Helper()
	raw, err := os.ReadFile(filepath.Join("..", "conformance.json"))
	if err != nil {
		t.Fatalf("read conformance.json: %v", err)
	}
	var vectors conformance
	if err := json.Unmarshal(raw, &vectors); err != nil {
		t.Fatalf("parse conformance.json: %v", err)
	}
	return vectors
}

func TestCategoryFoldMatchesSharedVectors(t *testing.T) {
	for _, vector := range loadConformance(t).CategoryFold {
		if got := CategoryKeyOf(vector.Raw); got != vector.Expect {
			t.Errorf("CategoryKeyOf(%q) = %q, want %q", vector.Raw, got, vector.Expect)
		}
	}
}

func TestResultShapeMatchesSharedVectors(t *testing.T) {
	for _, vector := range loadConformance(t).Results {
		result := BuildResult(true, vector.Detail)
		want := vector.Expect

		if result.Available != want.Available {
			t.Errorf("%s: Available = %v, want %v", vector.Name, result.Available, want.Available)
		}
		if result.Total != want.Total {
			t.Errorf("%s: Total = %d, want %d", vector.Name, result.Total, want.Total)
		}
		if result.ReportedTotal != want.ReportedTotal {
			t.Errorf("%s: ReportedTotal = %d, want %d", vector.Name, result.ReportedTotal, want.ReportedTotal)
		}
		for key, count := range want.Categories {
			if result.Categories[key] != count {
				t.Errorf("%s: Categories[%q] = %d, want %d", vector.Name, key, result.Categories[key], count)
			}
		}
		if result.IsAntibot() != want.IsAntibot {
			t.Errorf("%s: IsAntibot = %v, want %v", vector.Name, result.IsAntibot(), want.IsAntibot)
		}
		if result.IsProtected() != want.IsProtected {
			t.Errorf("%s: IsProtected = %v, want %v", vector.Name, result.IsProtected(), want.IsProtected)
		}
		if want.FirstName != "" {
			if result.Detections[0].Name != want.FirstName {
				t.Errorf("%s: first Name = %q, want %q", vector.Name, result.Detections[0].Name, want.FirstName)
			}
			if result.Detections[0].CategoryKey != want.FirstCategoryKey {
				t.Errorf("%s: first CategoryKey = %q, want %q",
					vector.Name, result.Detections[0].CategoryKey, want.FirstCategoryKey)
			}
		}
		if want.LastConfidenceIsNull {
			if last := result.Detections[len(result.Detections)-1]; last.Confidence != nil {
				t.Errorf("%s: a non-numeric confidence must be nil, got %v", vector.Name, *last.Confidence)
			}
		}
	}
}

func TestUnavailableMatchesSharedVectorAndIsNotClean(t *testing.T) {
	want := loadConformance(t).Unavailable.Expect
	result := Unavailable(ReasonUnavailable)

	if result.Available != want.Available {
		t.Errorf("Available = %v, want %v", result.Available, want.Available)
	}
	if result.Reason != want.Reason {
		t.Errorf("Reason = %q, want %q", result.Reason, want.Reason)
	}
	// false here must never be read as "clean" — Available says which it is.
	if result.IsAntibot() != want.IsAntibot || result.IsProtected() != want.IsProtected {
		t.Error("an unavailable result must not report detections")
	}
}

func TestEmbeddedBridgeMatchesCanonical(t *testing.T) {
	canonical, err := os.ReadFile(filepath.Join("..", "bridge.js"))
	if err != nil {
		t.Fatalf("read bridge.js: %v", err)
	}
	if strings.TrimSpace(bridgeScript) != strings.TrimSpace(string(canonical)) {
		t.Error("bridge.go has drifted from scrapeless-sdks/bridge.js")
	}
}

File: scrapeless-sdks/java/src/test/java/io/scrapeless/detector/ConformanceTest.java
package io.scrapeless.detector;

import java.nio.file.Files;
import java.nio.file.Path;
import java.util.List;
import java.util.Map;

/**
 * Reads ../../conformance.json, the shared truth for every SDK, so a fold that
 * drifts here fails here.
 *
 * <p>Plain assertions and a {@code main}, so it runs with nothing but a JDK:
 * {@code javac ... && java io.scrapeless.detector.ConformanceTest}.
 */
public final class ConformanceTest {
    private static int passed;
    private static int failed;

    private static void check(boolean condition, String label) {
        if (condition) {
            passed++;
            return;
        }
        failed++;
        System.err.println("  FAIL " + label);
    }

    @SuppressWarnings("unchecked")
    public static void main(String[] args) throws Exception {
        Path root = Path.of(args.length > 0 ? args[0] : "../..");
        Map<String, Object> vectors =
                (Map<String, Object>) MiniJson.parse(Files.readString(root.resolve("conformance.json")));

        for (Object entry : (List<Object>) vectors.get("categoryFold")) {
            Map<String, Object> vector = (Map<String, Object>) entry;
            String raw = (String) vector.get("raw");
            String expect = (String) vector.get("expect");
            check(Categories.keyOf(raw).equals(expect), "fold \"" + raw + "\" -> " + expect);
        }

        for (Object entry : (List<Object>) vectors.get("results")) {
            Map<String, Object> vector = (Map<String, Object>) entry;
            String name = (String) vector.get("name");
            Map<String, Object> want = (Map<String, Object>) vector.get("expect");
            DetectionResult result =
                    DetectionResult.build(true, (Map<String, Object>) vector.get("detail"));

            check(result.available() == (Boolean) want.get("available"), name + ": available");
            check(result.total() == ((Number) want.get("total")).intValue(), name + ": total");
            check(result.reportedTotal() == ((Number) want.get("reportedTotal")).intValue(),
                    name + ": reportedTotal");
            Map<String, Object> counts = (Map<String, Object>) want.get("categories");
            for (Map.Entry<String, Object> bucket : counts.entrySet()) {
                check(result.categories().get(bucket.getKey())
                                .equals(((Number) bucket.getValue()).intValue()),
                        name + ": categories." + bucket.getKey());
            }
            check(result.isAntibot() == (Boolean) want.get("isAntibot"), name + ": isAntibot");
            check(result.isProtected() == (Boolean) want.get("isProtected"), name + ": isProtected");

            if (want.get("firstName") instanceof String first) {
                check(result.detections().get(0).name().equals(first), name + ": firstName");
                check(result.detections().get(0).categoryKey().equals(want.get("firstCategoryKey")),
                        name + ": firstCategoryKey");
            }
            if (Boolean.TRUE.equals(want.get("lastConfidenceIsNull"))) {
                check(result.detections().get(result.detections().size() - 1).confidence() == null,
                        name + ": a non-numeric confidence must be null");
            }
        }

        Map<String, Object> want =
                (Map<String, Object>) ((Map<String, Object>) vectors.get("unavailable")).get("expect");
        DetectionResult missing = DetectionResult.unavailable(DetectionResult.REASON_UNAVAILABLE);
        check(missing.available() == (Boolean) want.get("available"), "unavailable: available");
        check(missing.reason().equals(want.get("reason")), "unavailable: reason");
        check(missing.total() == ((Number) want.get("total")).intValue(), "unavailable: total");
        // false here must never be read as "clean" — available says which it is.
        check(missing.isAntibot() == (Boolean) want.get("isAntibot"), "unavailable: isAntibot");
        check(missing.isProtected() == (Boolean) want.get("isProtected"), "unavailable: isProtected");

        String canonical = Files.readString(root.resolve("bridge.js")).trim();
        check(Bridge.INIT_SCRIPT.trim().equals(canonical),
                "Bridge.java matches scrapeless-sdks/bridge.js");

        Map<String, Object> sample =
                (Map<String, Object>) ((Map<String, Object>) ((List<Object>) vectors.get("results")).get(0))
                        .get("detail");
        DetectionResult result = DetectionResult.build(true, sample);
        check(result.ofCategory("antibot").size() == 1, "ofCategory filters to one bucket");
        check(result.ofCategory("antibot").get(0).name().equals("Cloudflare Bot Management"),
                "ofCategory returns the right one");

        Map<String, Object> kept = Map.of("detections",
                List.of(Map.of("name", "Akamai", "vendorNote", "keep me")));
        check("keep me".equals(DetectionResult.build(true, kept).detections().get(0).get("vendorNote")),
                "original fields survive");

        check(Categories.keyOfDetection(Map.of("detector", Map.of("category", "Anti-Bot")))
                .equals("antibot"), "nested detector shape");
        check(Categories.keyOfDetection(null).equals("other"), "null folds to other");

        System.out.printf("%n  %d passed, %d failed%n", passed, failed);
        System.exit(failed > 0 ? 1 : 0);
    }
}
